### PR TITLE
Feature refactor built-in coders and support animated webp on macOS

### PR DESCRIPTION
**Pull request by dreampiggy**
### New Pull Request Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar pull request in the [project](https://github.com/rs/SDWebImage/pulls) and found none

* [x] I have updated this branch with the latest master to avoid conflicts (via merge from master or rebase)
* [x] I have added the required tests to prove the fix/feature I am adding
* [x] I have updated the documentation (if necessary)
* [x] I have run the tests and they pass
* [x] I have run the lint and it passes (`pod lib lint`)

This merge request fixes / reffers to the following issues: ...

### Pull Request Description

macOS's NSImage API does not have something like UIImage's `animatedImageWithImages:duration:`. It only support GIF due to NSImage's API design and NSImageView implementation (Which take Image/IO source ref to decode data during rendering).

But I recently found a way to make it possible to let animated WebP work on macOS. We just grab the image frame array, then encoding it to a new GIF buffer and let NSImage load this temp GIF. Then all the animate tings work. (Mentioned here #2064). This is acceptable on macOS because it's more powerful and this happed under global queue from SD.

Yet I can do this by copying the encode GIF code from SDWebImageGIFCoder (because different coders should not relay on each other). But I dislike this way. And I decide to do some little refactor code which provide two extra method for animated images parse.

I create a new class `SDWebImageFrame`, which can be used for representing a frame for animated images. Since we can share code from that C function on `SDWebImageCoder`, but it's not really GOOD. So I create a new helper class called `SDWebImageCoderHelper`. I move all the trick logic previouslly in `SDWebImageWebPCoder` to this helper. Finally I add this `fake GIF` logic I talked above.

Meantime, I modify `NSImage+WebCache` that `sd_imageLoopCount` to automatically grab the GIF loop count because there are no way to set this previously. And I also fix one iOS 8-above API during previous plugin refactor. (And, any plan to upgrade minimum deployment version ? 😕 )

### CODE REVIEW

**Comment by dreampiggy** [comment]
A little more code 😅  Most code is just move from one place to another. Maybe you can revice through commits.

**Comment by bpoplauschi** [comment]
At first look, the code looks better. I'll have to review in more depth before I can say if there are no hidden issues :)
Anyway, since I just released 4.2.0, I've marked this PR for 5.0.0 release.

**Comment by dreampiggy** [comment]
@bpoplauschi You just relaesed 4.2.0 ? :) Wait...Even you do not merge this soon, please fix that iOS 8 API about `kCGImagePropertyOrientation ` because our minmum deployment target is still iOS 7.

Anyway, I think this is not a break change because it's a continue work after that coder refactor. Previously I move some code in that `SDWebImagCoder` global C function , which is not really good if you want to build your own coder. And since this changes only affect the build-in coder(The changes is not so different, just move code :) ). So I advice to add this one to 4.2.0 milestone. 

If we released this with that old `SDWebImageCoder` functions, I think it's not really easy to build your own one, isn't it 😅

**Comment by bpoplauschi** [comment]
@dreampiggy I asked you guys if there are any work remaining for 4.2.0 release. :) So it's gone already, released.
We can patch it. What is the problem with `kCGImagePropertyOrientation`?
If this PR is not breaking changes, we can release a 4.2.1 patch.

**Comment by dreampiggy** [comment]
@bpoplauschi `kCGImagePropertyOrientationUp` this is iOS 8 above API. Previously it's just a NSInteger number, but we missed that during refactor and change it to the enum. So this is the problem.

https://github.com/rs/SDWebImage/blob/master/SDWebImage/SDWebImageImageIOCoder.m#L521-587

**Comment by bpoplauschi** [comment]
I see

**Comment by dreampiggy** [comment]
This fix: https://github.com/dreampiggy/SDWebImage/blob/8f25f677dbd0a58747e6aca995ab276b31c3c61a/SDWebImage/SDWebImageCoderHelper.m#L179-L249

**Comment by dreampiggy** [comment]
@bpoplauschi This one: https://github.com/rs/SDWebImage/pull/2085

**Comment by dreampiggy** [comment]
@bpoplauschi I can change this to not break API since 4.2.0. Just leave previous C global function in `SDWebImageCoder` and move the new helper function to that `SDWebImageCoderHelper` class :). Could I mark this to 4.2.1 milestone ? (Also, @Pearapps may need this helper function to make their coder works better).

File: SDWebImage/NSImage+WebCache.m [hunk 1]
```diff
@@ -28,10 +28,8 @@ - (BOOL)isGIF {
         if ([rep isKindOfClass:[NSBitmapImageRep class]]) {
             NSBitmapImageRep *bitmapRep = (NSBitmapImageRep *)rep;
             NSUInteger frameCount = [[bitmapRep valueForProperty:NSImageFrameCount] unsignedIntegerValue];
-            if (frameCount > 1) {
-                isGIF = YES;
-                break;
-            }
+            isGIF = frameCount > 1 ? YES : NO;
```
Comment (dreampiggy) [inline SDWebImage/NSImage+WebCache.m]: NSImage 's design is that it use NSImageRep for actually rendering images. It choose the "best represent" to showing. In fact, GIF is bitmapRep without scale hint for PDF or someother things. So we do not need to check all image representations.

File: SDWebImage/SDWebImageGIFCoder.m [hunk 2]
```diff
@@ -38,43 +39,39 @@ - (UIImage *)decodedImageWithData:(NSData *)data {
...
         for (size_t i = 0; i < count; i++) {
-            CGImageRef image = CGImageSourceCreateImageAtIndex(source, i, NULL);
-            if (!image) {
+            CGImageRef imageRef = CGImageSourceCreateImageAtIndex(source, i, NULL);
+            if (!imageRef) {
                 continue;
             }
             
-            duration += [self sd_frameDurationAtIndex:i source:source];
+            float duration = [self sd_frameDurationAtIndex:i source:source];
```
Comment (dreampiggy) [inline SDWebImage/SDWebImageGIFCoder.m]: GIF use Image/IO. So it use seconds.

File: SDWebImage/SDWebImageImageIOCoder.m [hunk 3]
```diff
@@ -430,8 +431,8 @@ - (NSData *)encodedDataWithImage:(UIImage *)image format:(SDImageFormat)format {
     }
     
     NSMutableDictionary *properties = [NSMutableDictionary dictionary];
-#if SD_UIKIT
-    NSInteger exifOrientation = [[self class] sd_exifOrientationFromImageOrientation:image.imageOrientation];
+#if SD_UIKIT || SD_WATCH
```
Comment (dreampiggy) [inline SDWebImage/SDWebImageImageIOCoder.m]: This method is avilable for watchOS. So open it to solve potential orientation issue on watchOS.

File: SDWebImage/SDWebImageWebPCoder.m [hunk 4]
```diff
@@ -259,7 +230,7 @@ - (UIImage *)decompressedImageWithImage:(UIImage *)image
     return image;
 }
 
-- (nullable UIImage *)sd_blendWebpImageWithCanvas:(CGContextRef)canvas iterator:(WebPIterator)iter {
+- (nullable UIImage *)sd_drawnWebpImageWithCanvas:(CGContextRef)canvas iterator:(WebPIterator)iter {
```
Comment (dreampiggy) [inline SDWebImage/SDWebImageWebPCoder.m]: These two method is nearly the same except that blend mode process. For non-blend(Which means you should **cover** the current frame imaeg), just clear the target image rectangle before draw.

File: SDWebImage/SDWebImageWebPCoder.m [hunk 5]
```diff
@@ -379,23 +323,21 @@ - (NSData *)encodedDataWithImage:(UIImage *)image format:(SDImageFormat)format {
...
         if (!mux) {
             return nil;
         }
-        for (NSUInteger i = 0; i < images.count; i++) {
-            NSData *webpData = [self sd_encodedWebpDataWithImage:images[i]];
-            int duration = durations[i];
+        for (size_t i = 0; i < frames.count; i++) {
+            SDWebImageFrame *currentFrame = frames[i];
+            NSData *webpData = [self sd_encodedWebpDataWithImage:currentFrame.image];
+            int duration = currentFrame.duration * 1000;
```
Comment (dreampiggy) [inline SDWebImage/SDWebImageWebPCoder.m]: WebP encoder need milliseconds but not seconds. So don't forget to covert

File: SDWebImage/SDWebImageCoderHelper.m [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+
+#import "SDWebImageCoderHelper.h"
+#import "SDWebImageFrame.h"
+#import "UIImage+MultiFormat.h"
+#import "NSImage+WebCache.h"
+#import <ImageIO/ImageIO.h>
+
+@implementation SDWebImageCoderHelper
+
++ (UIImage *)animatedImageWithFrames:(NSArray<SDWebImageFrame *> *)frames {
```
Comment (dreampiggy) [inline SDWebImage/SDWebImageCoderHelper.m]: Copied from previous `SDWebImageWebPCoder` code. Use frame class instead of two arrays. :)

**Comment by dreampiggy** [comment]
@bpoplauschi Done. Revert the changes which breaking the public API. All new helper function is available in that `SDWebImageCoderHelper` class.

**Comment by bpoplauschi** [comment]
@dreampiggy I have reviewed this PR and everything looks good (I've added a small commit just making the new classes public for people who need to write custom coders). I will merge this into master (actually squash) and then release a 4.2.1 patch. This means I will close #2085.

**Comment by bpoplauschi** [comment]
4.2.1 is released.

**Comment by bpoplauschi** [comment]
@dreampiggy I really appreciate your dedication to this project. Good job on this refactor

**Pull request merged**